vizro-core/tests/unit/vizro/models/_action/test_action.py

```diff
@@ -0,0 +1,246 @@
+        assert action_components == []
+
+    @pytest.mark.parametrize(
+        "custom_action_function_mock_return, callback_context_set_outputs_grouping, expected_function_return_value",
+        [
+            # custom action function return value - None
+            (None, [], {"action_finished": None}),
+            # custom action function return value - single value
+            ("new_value", ["component_1_property"], {"action_finished": None, "component_1_property": "new_value"}),
+            # custom action function return value - list of values
+            (
+                ["new_value", "new_value_2"],
+                ["component_1_property", "component_2_property"],
+                {"action_finished": None, "component_1_property": "new_value", "component_2_property": "new_value_2"},
+            ),
+            # custom action function return value - tuple
+            (
+                ("new_value", "new_value_2"),
+                ["component_1_property", "component_2_property"],
+                {"action_finished": None, "component_1_property": "new_value", "component_2_property": "new_value_2"},
+            ),
+            # custom action function return value - dictionary
+            (
+                {"component_1_property": "new_value"},
+                ["component_1_property"],
+                {"action_finished": None, "component_1_property": "new_value"},
+            ),
+        ],
+        indirect=["custom_action_function_mock_return", "callback_context_set_outputs_grouping"],
+    )
+    def test_action_callback_function_return_value_valid(
+        self, custom_action_function_mock_return, callback_context_set_outputs_grouping, expected_function_return_value
+    ):
+        action = Action(function=custom_action_function_mock_return())
+        result = action._action_callback_function()
+        assert result == expected_function_return_value
+
+    @pytest.mark.parametrize(
+        "custom_action_function_mock_return, callback_context_set_outputs_grouping",
+        [
+            (None, ["component_1_property"]),
+            ("new_value", []),
+            ("new_value", ["component_1_property", "component_2_property"]),
+            (["new_value"], []),
+            (["new_value"], ["component_1_property", "component_2_property"]),
+            (["new_value", "new_value_2"], ["component_1_property"]),
+            ({"component_1_property": "new_value"}, []),
+            ({"component_1_property": "new_value"}, ["component_1_property", "component_2_property"]),
+            ({"component_1_property": "new_value", "component_2_property": "new_value_2"}, ["component_1_property"]),
+        ],
+        indirect=True,
+    )
+    def test_action_callback_function_return_value_invalid(
+        self, custom_action_function_mock_return, callback_context_set_outputs_grouping
+    ):
+        action = Action(function=custom_action_function_mock_return())
+        with pytest.raises(
+            ValueError,
```

Comment:
Could this be a `ValidationError`?

Reply:
How do you mean? To raise a `pydantic.ValidationError`?
It doesn't seem that worthwhile to me because the  `pydantic.ValidationError` requires a Model in the constructor. 
In that case I'd need to create a new model (e.g. LengthValidatior(element_1, element_2)) that has two arguments, and a validator that asserts `assert len(element_1) == len(element_2)`.

Reply:
Ah yes, sorry - I misread the test. I thought the `Action()` was inside the `pytest.raises` so would be raising `ValidationError`. Ignore me.